File: android/engine/bin/main/io/sauti/engine/Mesh.kt
package io.sauti.engine

import kotlinx.coroutines.CoroutineScope

interface MeshHooks {
    fun send(text: String)
    fun onConnected(peerId: String)
    fun onRemoteAudio(peerId: String)
    fun onError(error: Throwable)
}

class Mesh(
    private val scope: CoroutineScope,
    private val factory: RtcFactory,
    private val selfId: String,
    private val hooks: MeshHooks
) {
    private val peers = LinkedHashMap<String, Peer>()
    private var iceServers: List<IceServer> = emptyList()

    private val signal = object : PeerSignal {
        override fun send(text: String) = hooks.send(text)
        override fun onRemoteAudio(peerId: String) = hooks.onRemoteAudio(peerId)
        override fun onUnrecoverable(peerId: String) = rebuild(peerId)
        override fun onConnected(peerId: String) = hooks.onConnected(peerId)
        override fun onError(error: Throwable) = hooks.onError(error)
    }

    fun setIceServers(servers: List<IceServer>) {
        iceServers = servers
    }

    fun has(peerId: String): Boolean = peers.containsKey(peerId)

    fun ids(): List<String> = peers.keys.toList()

    fun count(): Int = peers.size

    fun ensurePeer(peerId: String): Peer = peers[peerId] ?: createPeer(peerId)

    private fun createPeer(peerId: String): Peer {
        val pc = factory.createPeerConnection(iceServers)
        val peer = Peer(peerId, selfId, pc, scope, signal)
        peers[peerId] = peer
        return peer
    }

    fun addPeer(peerId: String): Peer {
        val peer = createPeer(peerId)
        peer.start()
        return peer
    }

    private fun rebuild(peerId: String) {
        val old = peers[peerId] ?: return
        old.close()
        peers.remove(peerId)
        createPeer(peerId).start()
    }

    fun removePeer(peerId: String) {
        val peer = peers[peerId] ?: return
        peer.close()
        peers.remove(peerId)
    }

    suspend fun routeOffer(from: String, sdp: String) {
        ensurePeer(from).onOffer(sdp)
    }

    suspend fun routeAnswer(from: String, sdp: String) {
        peers[from]?.onAnswer(sdp)
    }

    suspend fun routeIce(from: String, candidate: IceCandidate) {
        peers[from]?.onIce(candidate)
    }

    fun iceRestartAll() {
        for (peer in peers.values) peer.iceRestart()
    }

    fun statsPeer(peerId: String): Peer? = peers[peerId]

    fun teardown() {
        for (peer in peers.values) peer.close()
        peers.clear()
    }
}

File: packages/core/src/mesh.ts
import type { AudioSinkManager } from './audio.js';
import { Peer, type PeerSignal } from './peer.js';
import type { OpusOptions } from './sdp.js';
import type {
  IceCandidateLike,
  IceServerConfig,
  SautiRuntime,
  StreamLike,
  TrackLike
} from './types.js';

export interface MeshHooks {
  send(frame: unknown): void;
  onConnected(peerId: string): void;
  onError(error: unknown): void;
}

export class Mesh {
  private readonly peers = new Map<string, Peer>();
  private iceServers: IceServerConfig[] = [];
  private local: { stream: StreamLike; track: TrackLike } | null = null;

  private readonly signal: PeerSignal;

  constructor(
    private readonly runtime: SautiRuntime,
    private readonly sinks: AudioSinkManager,
    private readonly selfId: string,
    private readonly hooks: MeshHooks,
    private readonly opus: OpusOptions = {}
  ) {
    this.signal = {
      send: (frame) => this.hooks.send(frame),
      onRemoteStream: (peerId, stream) => this.sinks.attach(peerId, stream),
      onUnrecoverable: (peerId) => this.rebuild(peerId),
      onConnected: (peerId) => this.hooks.onConnected(peerId),
      onError: (error) => this.hooks.onError(error)
    };
  }

  setIceServers(servers: IceServerConfig[]): void {
    this.iceServers = servers;
  }

  setLocalMedia(stream: StreamLike, track: TrackLike): void {
    this.local = { stream, track };
  }

  has(peerId: string): boolean {
    return this.peers.has(peerId);
  }

  ids(): string[] {
    return [...this.peers.keys()];
  }

  ensurePeer(peerId: string): Peer {
    const existing = this.peers.get(peerId);
    if (existing) return existing;
    return this.createPeer(peerId);
  }

  private createPeer(peerId: string): Peer {
    if (!this.local) {
      throw new Error('local media not initialized');
    }
    const pc = this.runtime.createPeerConnection({ iceServers: this.iceServers });
    const peer = new Peer(
      peerId,
      this.selfId,
      pc,
      this.local.stream,
      this.local.track,
      this.signal,
      this.opus
    );
    this.peers.set(peerId, peer);
    return peer;
  }

  addPeer(peerId: string): Peer {
    const peer = this.createPeer(peerId);
    peer.start();
    return peer;
  }

  private rebuild(peerId: string): void {
    const old = this.peers.get(peerId);
    if (!old) return;
    old.close();
    this.peers.delete(peerId);
    const peer = this.createPeer(peerId);
    peer.start();
  }

  removePeer(peerId: string): void {
    const peer = this.peers.get(peerId);
    if (!peer) return;
    peer.close();
    this.peers.delete(peerId);
    this.sinks.remove(peerId);
  }

  async routeOffer(from: string, sdp: string): Promise<void> {
    const peer = this.ensurePeer(from);
    await peer.onOffer(sdp);
  }

  async routeAnswer(from: string, sdp: string): Promise<void> {
    const peer = this.peers.get(from);
    if (peer) await peer.onAnswer(sdp);
  }

  async routeIce(from: string, candidate: IceCandidateLike): Promise<void> {
    const peer = this.peers.get(from);
    if (peer) await peer.onIce(candidate);
  }

  iceRestartAll(): void {
    for (const peer of this.peers.values()) peer.iceRestart();
  }

  async replaceTrackAll(track: TrackLike): Promise<void> {
    if (this.local) this.local.track = track;
    for (const peer of this.peers.values()) await peer.replaceTrack(track);
  }

  teardown(): void {
    for (const peer of this.peers.values()) peer.close();
    this.peers.clear();
    this.sinks.clear();
  }
}
